File: python/netcheck.py
#!/usr/bin/env python3
"""Confirms san_db_ox._codec stays free of networking/subprocess imports
(.claude/rules/architecture.md: "コーデック層はネットワークに依存しない
ことを検証可能にする"), the same idea as san-db-ox's own `make netcheck`
and this repo's `make go-netcheck`.

Importing san_db_ox._codec the naive way (`import san_db_ox._codec`) would
first run san_db_ox/__init__.py, which re-exports the client and pulls in
subprocess/socket transitively -- so this script loads _codec.py in
isolation via a synthetic package instead, sidestepping __init__.py
entirely.
"""

from __future__ import annotations

import ast
import importlib
import pathlib
import sys
import types

SRC = pathlib.Path(__file__).parent / "src" / "san_db_ox"
CODEC_FILE = SRC / "_codec.py"

# Deliberately small and explicit: any new top-level import in _codec.py
# must be a conscious decision (and an update to this set), not something
# that slips in unnoticed during a refactor.
ALLOWED_ROOTS = {
    "__future__",
    "base64",
    "binascii",
    "collections",
    "dataclasses",
    "json",
    "math",
    "typing",
}

# Checked independently of ALLOWED_ROOTS above: even if _codec.py's own
# import statements look fine, something they import could transitively
# pull one of these in.
FORBIDDEN_ROOTS = {
    "socket",
    "_socket",
    "ssl",
    "_ssl",
    "subprocess",
    "_posixsubprocess",
    "http",
    "urllib",
    "asyncio",
    "selectors",
    "multiprocessing",
}


def check_static_imports() -> list[str]:
    tree = ast.parse(CODEC_FILE.read_text(encoding="utf-8"), filename=str(CODEC_FILE))
    bad: list[str] = []
    for node in ast.walk(tree):
        if isinstance(node, ast.Import):
            for alias in node.names:
                root = alias.name.split(".")[0]
                if root not in ALLOWED_ROOTS:
                    bad.append(f"import {alias.name} (line {node.lineno})")
        elif isinstance(node, ast.ImportFrom):
            if node.module is None:  # relative "from . import x"
                continue
            root = node.module.split(".")[0]
            if root not in ALLOWED_ROOTS:
                bad.append(f"from {node.module} import ... (line {node.lineno})")
    return bad


def check_dynamic_imports() -> list[str]:
    # Only modules newly loaded by importing _codec count -- this script's
    # own top-level imports (e.g. pathlib, which drags in urllib.parse for
    # as_uri()) are already in sys.modules beforehand and aren't _codec's
    # doing.
    before = set(sys.modules)

    pkg_name = "_san_db_ox_netcheck_probe"
    pkg = types.ModuleType(pkg_name)
    pkg.__path__ = [str(SRC)]
    sys.modules[pkg_name] = pkg
    importlib.import_module(f"{pkg_name}._codec")

    newly_loaded = set(sys.modules) - before
    return sorted(
        name for name in newly_loaded if name.split(".")[0] in FORBIDDEN_ROOTS
    )


def main() -> int:
    problems: list[str] = []

    bad_imports = check_static_imports()
    if bad_imports:
        problems.append(
            "_codec.py imports modules outside the allowed set "
            f"({sorted(ALLOWED_ROOTS)}):\n  " + "\n  ".join(bad_imports)
        )

    bad_modules = check_dynamic_imports()
    if bad_modules:
        problems.append(
            "loading san_db_ox._codec in isolation pulled in forbidden "
            f"modules: {bad_modules}"
        )

    if problems:
        print("python-netcheck FAILED:\n\n" + "\n\n".join(problems), file=sys.stderr)
        return 1
    print("python-netcheck: san_db_ox._codec is free of networking/subprocess imports")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: typescript/netcheck.mjs
#!/usr/bin/env node
// Mechanically confirms src/codec.ts stays free of networking/process
// machinery (.claude/rules/architecture.md: "コーデック層はネットワークに
// 依存しないことを検証可能にする"), the same idea as go-netcheck (`go list
// -deps | grep`) and python/netcheck.py (an import allowlist plus a
// sys.modules diff after loading _codec.py in isolation).
//
// Plain .mjs, run directly by `node netcheck.mjs` with no build step of its
// own -- it inspects the already-built dist/codec.js (via
// `make typescript-netcheck: typescript-build`), which is what actually
// ships.

import { readFile } from "node:fs/promises";
import { fileURLToPath } from "node:url";
import ts from "typescript";

const CODEC_SRC = fileURLToPath(new URL("./src/codec.ts", import.meta.url));
const CODEC_DIST = fileURLToPath(new URL("./dist/codec.js", import.meta.url));

// Anything reachable through these substrings, transitively, means the
// codec layer pulled in networking, TLS, HTTP, or child-process machinery.
const FORBIDDEN_SUBSTRINGS = [
  "net",
  "tls",
  "http",
  "http2",
  "https",
  "dgram",
  "dns",
  "child_process",
  "cluster",
  "worker_threads",
  "inspector",
  "repl",
  "fs",
  "tcp_wrap",
  "pipe_wrap",
  "udp_wrap",
  "spawn_sync",
  "process_wrap",
  "cares_wrap",
];

function isForbidden(moduleLoadListEntry) {
  // Entries look like "NativeModule net" or "Internal Binding tcp_wrap".
  const name = moduleLoadListEntry.split(" ").at(-1) ?? "";
  return FORBIDDEN_SUBSTRINGS.some(
    (bad) =>
      name === bad || name.startsWith(`${bad}/`) || name.startsWith(`node:${bad}`),
  );
}

let failed = false;
function fail(message) {
  failed = true;
  console.error(`typescript-netcheck FAILED: ${message}`);
}

// -- (a) Static import allowlist --------------------------------------------
//
// The codec's intended import count is zero: base64 uses the Buffer
// global, encoding uses TextEncoder/TextDecoder globals, none of which are
// imports. Any import at all must be relative (a sibling module in this
// driver, e.g. a future split of codec.ts) -- a bare specifier ("node:*" or
// an npm package) is disqualifying on its own.
{
  const source = await readFile(CODEC_SRC, "utf-8");
  const info = ts.preProcessFile(
    source,
    /* readImportFiles */ true,
    /* detectJavaScriptImports */ true,
  );
  const bad = info.importedFiles
    .map((f) => f.fileName)
    .filter((spec) => !spec.startsWith("."));
  if (bad.length > 0) {
    fail(`src/codec.ts has non-relative import(s): ${JSON.stringify(bad)}`);
  } else {
    console.log(
      `static check: src/codec.ts imports ${info.importedFiles.length} module(s), all relative.`,
    );
  }
}

// -- (b)/(c) Runtime module-load-graph diff, with a positive control --------
//
// Runs in a fresh child `node -e` so the probe's own module graph doesn't
// contaminate the measurement, and so the positive control (importing
// node:net) doesn't leave net loaded for the real probe.
async function loadListDiff(specifier) {
  const { execFileSync } = await import("node:child_process");
  const script = `
    const before = new Set(process.moduleLoadList);
    await import(${JSON.stringify(specifier)});
    const diff = process.moduleLoadList.filter((m) => !before.has(m));
    process.stdout.write(JSON.stringify(diff));
  `;
  const out = execFileSync(process.execPath, ["--input-type=module", "-e", script], {
    encoding: "utf-8",
  });
  return JSON.parse(out);
}

{
  // Positive control first: process.moduleLoadList is undocumented, so
  // prove the detection mechanism itself still works before trusting a
  // clean result from codec.js. If Node ever stops reporting builtin loads
  // this way, this must fail loudly rather than let the real check pass by
  // accident.
  const controlDiff = await loadListDiff("node:net");
  if (!controlDiff.some(isForbidden)) {
    fail(
      "netcheck's detection mechanism is broken (process.moduleLoadList no " +
        "longer reports loaded builtins for node:net) -- fix netcheck.mjs " +
        "before trusting it",
    );
  } else {
    console.log("positive control: importing node:net is detected, as expected.");
  }

  const codecDiff = await loadListDiff(CODEC_DIST);
  const codecBad = codecDiff.filter(isForbidden);
  if (codecBad.length > 0) {
    fail(
      `dist/codec.js transitively loaded forbidden module(s): ${JSON.stringify(codecBad)}`,
    );
  } else {
    console.log(
      `runtime check: importing dist/codec.js loaded ${codecDiff.length} module(s), none forbidden.`,
    );
  }
}

if (failed) {
  process.exit(1);
}
console.log("typescript-netcheck OK");
